src/Sundials.jl

```diff
@@ -49,17 +52,80 @@ include("constants.jl")
 
 ##################################################################
 #
+# Types to facilitate release of memory allocated by the library
+#
+##################################################################
+type KINSOL   # memory handle for KINSOL
+    kinsol::Vector{KINSOL_ptr} # vector for passing to functions expecting Ptr{KINSOL_ptr}
+    function KINSOL()
+        k = new([KINCreate()])
+        finalizer(k, KINFree)
+        return k
+    end
+end
+Base.convert(::Type{KINSOL_ptr}, k::KINSOL) = k.kinsol[1]
+Base.convert(T::Type{Ptr{KINSOL_ptr}}, k::KINSOL) = convert(T, k.kinsol)
+
+type CVODE    # memory handle for CVode
+    cvode::Vector{CVODE_ptr} # vector for passing to functions expecting Ptr{CVODE_ptr}
+    function CVODE(lmm::Int, iter::Int)
+        k = new([CVodeCreate(lmm, iter)])
+        finalizer(k, CVodeFree)
+        return k
+    end
+end
+Base.convert(::Type{CVODE_ptr}, k::CVODE) = k.cvode[1]
+Base.convert(T::Type{Ptr{CVODE_ptr}}, k::CVODE) = convert(T, k.cvode)
+
+type IDA # memory handle for IDA
+    ida::Vector{IDA_ptr} # vector for passing to functions expecting Ptr{IDA_ptr}
+    function IDA()
+        k = new([IDACreate()])
+        finalizer(k, IDAFree)
+        return k
+    end
+end
+Base.convert(::Type{IDA_ptr}, k::IDA) = k.ida[1]
+Base.convert(T::Type{Ptr{IDA_ptr}}, k::IDA) = convert(T, k.ida)
+
+type NVector <: DenseVector{RealType} # memory handle for NVectors
+    ptr::Vector{N_Vector} # vector for passing to functions expecting Ptr{N_Vector}
+    v::Vector{RealType}
+
+    function NVector(x::Vector{RealType})
+        k = new([N_VMake_Serial(length(x), x)], x)
+        finalizer(k, N_VDestroy_Serial)
+        return k
+    end
+end
+NVector{T<:Real}(x::Vector{T}) = NVector(copy!(similar(x, RealType), x))
+
```

Comment:
Should probably also define
```jl
Base.similar(A::NVector, T=eltype(A), dims=size(A)) = NVector(similar(A.v))
```
so that e.g. `copy` works, and probably:
```jl
if VERSION < v"0.4.0-dev+3710"
    Base.convert(T::Type{Ptr{RealType}}, v::NVector) = convert(T, v.v)
else
    Base.unsafe_convert(T::Type{Ptr{RealType}}, v::NVector) = unsafe_convert(T, v.v)
end
```